common/cuda_hip/factorization/lu_kernels.hpp.inc

```diff
@@ -0,0 +1,206 @@
+        const auto dep_local = dep % (default_block_size / config::warp_size);
+        auto val = vals[lower_nz];
+        const auto diag_idx = diag_idxs[dep];
+        const auto dep_end = row_ptrs[dep + 1];
+        assert(dep < row);
+        if (dep_block == block) {
+            // wait for a local dependency
+            while (!load(sh_ready, dep_local)) {
+                __threadfence();
+            }
+        } else {
+            // wait for a global dependency
+            while (!load(ready, dep)) {
+                __threadfence();
+            }
+        }
+        __threadfence();
+        const auto diag = vals[diag_idx];
+        const auto scale = val / diag;
+        if (lane == 0) {
+            vals[lower_nz] = scale;
+        }
+        // subtract all entries past the diagonal
+        for (auto upper_nz = diag_idx + 1 + lane; upper_nz < dep_end;
+             upper_nz += config::warp_size) {
+            const auto upper_col = cols[upper_nz];
+            const auto upper_val = vals[upper_nz];
+            const auto output_pos = lookup.lookup_unsafe(upper_col) + row_begin;
+            vals[output_pos] -= scale * upper_val;
+        }
+    }
+    __threadfence();
+    // notify local warps
+    store(sh_ready, row_local, 1);
+    // notify other blocks
+    store(ready, row, 1);
+}
+
+
+}  // namespace kernel
+
+
+template <typename ValueType, typename IndexType>
+void initialize(std::shared_ptr<const DefaultExecutor> exec,
+                const matrix::Csr<ValueType, IndexType>* mtx,
+                const IndexType* factor_lookup_offsets,
+                const int64* factor_lookup_descs,
+                const int32* factor_lookup_storage, IndexType* diag_idxs,
+                matrix::Csr<ValueType, IndexType>* factors)
+{
+    const auto num_rows = mtx->get_size()[0];
+    if (num_rows > 0) {
+        const auto num_blocks =
+            ceildiv(num_rows, default_block_size / config::warp_size);
+        kernel::initialize<<<num_blocks, default_block_size>>>(
+            mtx->get_const_row_ptrs(), mtx->get_const_col_idxs(),
+            as_device_type(mtx->get_const_values()),
+            factors->get_const_row_ptrs(), factors->get_const_col_idxs(),
+            factor_lookup_offsets, factor_lookup_storage, factor_lookup_descs,
+            as_device_type(factors->get_values()), diag_idxs, num_rows);
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(GKO_DECLARE_LU_INITIALIZE);
+
+
+template <typename ValueType, typename IndexType>
+void factorize(std::shared_ptr<const DefaultExecutor> exec,
+               const IndexType* lookup_offsets, const int64* lookup_descs,
+               const int32* lookup_storage, const IndexType* diag_idxs,
+               matrix::Csr<ValueType, IndexType>* factors,
+               array<int>& tmp_storage)
+{
+    const auto num_rows = factors->get_size()[0];
+    if (num_rows > 0) {
+        tmp_storage.resize_and_reset(num_rows + 1);
```

Comment:
isn't `tmp_storage.resize_and_reset(2)` enough? You are only using the first and last entry, and ignoring everything in between.

Reply:
we use tmp_storage for two things: the atomic counter and 0/1 flags for each row

Reply:
Oh, I see, I didn't realize that the `store` does the indexing internally.